File: dataModule/src/main/java/com/mobnetic/coinguardian/model/market/SurBTC.java
package com.mobnetic.coinguardian.model.market;

import com.mobnetic.coinguardian.model.CheckerInfo;
import com.mobnetic.coinguardian.model.CurrencyPairInfo;
import com.mobnetic.coinguardian.model.Market;
import com.mobnetic.coinguardian.model.Ticker;
import com.mobnetic.coinguardian.model.currency.Currency;
import com.mobnetic.coinguardian.model.currency.VirtualCurrency;
import org.json.JSONArray;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;

/**
 * Implements SurBTC connection to the API
 * @author Eduardo Laguna
 */
public class SurBTC extends Market {

    private final static String NAME = "SurBTC";
    private final static String TTS_NAME = "sur BTC";
    private final static String URL = "https://www.surbtc.com/api/v2/markets/%1$s-%2$s/ticker";
    private final static String URL_CURRENCY_PAIRS = "https://www.surbtc.com/api/v2/markets.json";
    private final static HashMap<String, CharSequence[]> CURRENCY_PAIRS = new LinkedHashMap<String, CharSequence[]>();

    public SurBTC() {
        super(NAME, TTS_NAME, null);
    }

    @Override
    public String getUrl(int requestId, CheckerInfo checkerInfo) {
        return String.format(URL, checkerInfo.getCurrencyBaseLowerCase(), checkerInfo.getCurrencyCounterLowerCase());
    }

    @Override
    public String getCurrencyPairsUrl(int requestId) {
        return URL_CURRENCY_PAIRS;
    }

    @Override
    protected void parseCurrencyPairsFromJsonObject(int requestId, JSONObject jsonObject, List<CurrencyPairInfo> pairs) throws Exception {
        JSONArray markets = jsonObject.getJSONArray("markets");
        for (int i = 0; i < markets.length(); i++) {
            JSONObject market = markets.getJSONObject(i);
            pairs.add(new CurrencyPairInfo(market.getString("base_currency"), market.getString("quote_currency"), market.getString("id")));
        }
        super.parseCurrencyPairsFromJsonObject(requestId, jsonObject, pairs);
    }

    @Override
    protected void parseTickerFromJsonObject(int requestId, JSONObject jsonObject, Ticker ticker, CheckerInfo checkerInfo) throws Exception {
        JSONObject jsonTicker = jsonObject.getJSONObject("ticker");
        ticker.bid = jsonTicker.getJSONArray("max_bid").getDouble(0);
        ticker.ask = jsonTicker.getJSONArray("min_ask").getDouble(0);
        ticker.vol = jsonTicker.getJSONArray("volume").getDouble(0);
        ticker.last = jsonTicker.getJSONArray("last_price").getDouble(0);
    }
}

File: dataModule/src/main/java/com/aneonex/bitcoinchecker/datamodule/model/market/SurBtc.kt
package com.aneonex.bitcoinchecker.datamodule.model.market

import com.aneonex.bitcoinchecker.datamodule.model.CheckerInfo
import com.aneonex.bitcoinchecker.datamodule.model.CurrencyPairInfo
import com.aneonex.bitcoinchecker.datamodule.model.Market
import com.aneonex.bitcoinchecker.datamodule.model.Ticker
import org.json.JSONObject

/**
 * Implements SurBTC connection to the API
 * @author Eduardo Laguna
 */
class SurBtc : Market(NAME, TTS_NAME, null) {
    override fun getUrl(requestId: Int, checkerInfo: CheckerInfo): String {
        return String.format(URL, checkerInfo.currencyPairId)
    }

    @Throws(Exception::class)
    override fun parseTickerFromJsonObject(requestId: Int, jsonObject: JSONObject, ticker: Ticker, checkerInfo: CheckerInfo) {
        val tickerJsonObject = jsonObject.getJSONObject("ticker")
        ticker.bid = tickerJsonObject.getJSONArray("max_bid").getDouble(0)
        ticker.ask = tickerJsonObject.getJSONArray("min_ask").getDouble(0)
        ticker.vol = tickerJsonObject.getJSONArray("volume").getDouble(0)
        ticker.last = tickerJsonObject.getJSONArray("last_price").getDouble(0)
    }

    // ====================
    // Get currency pairs
    // ====================
    override fun getCurrencyPairsUrl(requestId: Int): String? {
        return URL_CURRENCY_PAIRS
    }

    @Throws(Exception::class)
    override fun parseCurrencyPairsFromJsonObject(requestId: Int, jsonObject: JSONObject, pairs: MutableList<CurrencyPairInfo>) {
        val marketsJsonArray = jsonObject.getJSONArray("markets")
        for (i in 0 until marketsJsonArray.length()) {
            val marketJsonObject = marketsJsonArray.getJSONObject(i)
            pairs.add(CurrencyPairInfo(
                    marketJsonObject.getString("base_currency"),
                    marketJsonObject.getString("quote_currency"),
                    marketJsonObject.getString("id")))
        }
    }

    companion object {
        private const val NAME = "SurBtc"
        private const val TTS_NAME = "Sur BTC"
        private const val URL = "https://www.surbtc.com/api/v2/markets/%1\$s/ticker.json"
        private const val URL_CURRENCY_PAIRS = "https://www.surbtc.com/api/v2/markets.json"
    }
}